.github/workflows/tests.yml

```diff
@@ -1,46 +1,222 @@
+    - name: Checkout repository
+      uses: actions/checkout@v2
     - name: Set up Python ${{ matrix.python-version }}
       uses: actions/setup-python@v2
       with:
         python-version: ${{ matrix.python-version }}
-    - name: Install dependencies
+    - name: Export ${HOME}/.local/bin to ${PATH}
+      # Executable Python binaries are ususally stored there
+      run: echo "${HOME}/.local/bin" >> ${GITHUB_PATH}
+    - name: Get pip cache dir
+      # pip's cache path depends on the operating system, see
+      # https://github.com/actions/cache/blob/main/examples.md#python---pip
+      # This requires pip >=20.1
+      id: pip-cache
       run: |
-        python -m pip install --upgrade pip
-        python -m pip install flake8 flake8-docstrings pytest
-        python -m pip install numpy
-        if [ -f requirements.txt ]; then pip install -r requirements.txt; fi
-    - name: Lint code with flake8
+        python -m pip install --user --upgrade pip
+        echo "::set-output name=dir::$(pip cache dir)"
+    - name: Create/Restore Cache
+      uses: actions/cache@v2
+      with:
+        path: |
+          ${{ steps.pip-cache.outputs.dir }}/**
+          ./docs/build/**
+          ~/docs/build/**
+          ./.pytest_cache/**
+          ~/.pytest_cache/**
+        key: ${{ runner.os }}-${{ matrix.python-version }}-${{ github.job }}
+        restore-keys: |
+          ${{ runner.os }}-${{ matrix.python-version }}-
+          ${{ runner.os }}-
+    - name: Install/Upgrade setuptools and wheel
+      run: python -m pip install --user --upgrade setuptools wheel
+    - name: Test installation of MDTools
       run: |
-        # stop the build if there are Python syntax errors or undefined names
-        flake8 . --count --select=E1,E2,E3,E4,E9,F63,F4,F7,F82,W1,W2,W3 --show-source --statistics
-        # exit-zero treats all errors as warnings
-        flake8 . --count --exit-zero --max-complexity=10 --max-line-length=79 --statistics
-    - name: Lint docstrings with flake8
+        python -m pip install --user --upgrade .
+        python -c "import mdtools"
+    - name: Install/Upgrade requirements to build the documentation
+      run: python -m pip install --user --upgrade -r docs/requirements.txt
+    - name: Test build of the documentation
+      run: make -C ./docs/ html
+    - name: doctest
+      # Because the default dtypes of NumPy arrays are different on
+      # Windows, doctest will detect false failures. See e.g.
+      # https://stackoverflow.com/questions/36278590/numpy-array-dtype-is-coming-as-int32-by-default-in-a-windows-10-64-bit-machine
+      if: ${{ runner.os != 'Windows' }}
+      run: make -C ./docs/ doctest
+    - name: Install/Upgrade PyTest
+      run: python -m pip install --user --upgrade pytest
+    - name: Test code with pytest
+      run: python -m pytest ./
+
+  # Scan the code for security vulnerabilities with CodeQL.
+  # See: https://github.com/github/codeql-action
+  codeql:
```

Comment:
This seems very cool. Although possibly a little bit overkill for a collection of scientific analysis scripts. But let's keep it and see how it works.

Reply:
True, but it improves the quality of the project :D

Reply:
I'm curious what this finds in the future.